app/src/main/java/com/duckduckgo/app/browser/favicon/FaviconPersister.kt

```diff
@@ -61,46 +63,41 @@ class FileBasedFaviconPersister(
         }
     }
 
-    override suspend fun store(directory: String, subFolder: String, bitmap: Bitmap, domain: String): File {
-        return withContext(dispatcherProvider.io()) {
-
+    override suspend fun store(directory: String, subFolder: String, bitmap: Bitmap, domain: String): File? {
+        return runBlocking(dispatcherProvider.io() + NonCancellable) {
```

Comment:
I don't think this is quite right. we shouldn't need to use `runBlocking` inside of a `suspend` function; using `withContext` will already wait until the `store` function has finished before return execution. Also, if this was called from different threads, you could still have multiple threads writing at the same time.

Is part of the problem around race conditions that `dispatcherProvider.io()` returns a dispatcher with a thread pool? so that even though you are using the same dispatcher each call, it might be delegating work to a different thread each time and so you'd potentially have multiple threads writing at the same time?

If the problem is limited only to the `store` function, would marking that as `synchronized` help any?

Reply:
Yeah that seemed to be the problem, `withContext` didn't wait at all and adding `runBlocking` did the trick. The whole idea was to synchronously execute the store method. 

I've changed the code a bit and extracted the store logic after `withContext` into a separate method and annotated with `@Synchronized`. Seems to work fine.